n_beats/model_keras.py

```diff
@@ -0,0 +1,296 @@
+                    backcast = Lambda(trend_model, arguments={"is_forecast": False, "length": self.backcast_length})(
+                        theta_b)
+                    forecast = Lambda(trend_model, arguments={"is_forecast": True, "length": self.forecast_length})(
+                        theta_f)
+                    x_ = Subtract()([x_, backcast])
+                    if i == 0 and j == 0:
+                        y_ = forecast
+                    else :
+                        y_ = Add()([y_, forecast])
+                if self.block_types[j] == 'seasonality':
+                    theta_b = theta_f = Dense(self.backcast_length, activation='linear')(d4)
+                    # thetas_b = Dense(self.backcast_length, activation='linear')(d4) 
+                    # theta_b and theta_f are shared even for seasonality bloc
+                    backcast = Lambda(seasonality_model,
+                                      arguments={"is_forecast": False, "length": self.backcast_length})(theta_b)
+                    forecast = Lambda(seasonality_model,
+                                      arguments={"is_forecast": True, "length": self.forecast_length})(theta_f)
+                    x_ = Subtract()([x_, backcast])
+                    if i == 0 and j == 0:
+                        y_ = forecast
+                    else :
+                        y_ = Add()([y_, forecast])
+
+        model = Model(X_, y_)
+        model.summary()
+
+        optimizer = Adam(lr=0.00001)
+        self.nbeats = model
+        self.nbeats.compile(loss='mae', optimizer=optimizer)
+
+    def train_model(self):
+
+        print('loading dev data')
+        x_test, y_test = get_xtest_circular(self.backcast_length, self.forecast_length)
+        print(x_test.shape)
+
+        nb_epochs = self.epochs
+        for step in range(nb_epochs):
+            # print('loading train data')
+            x, y = get_xtrain_circular(self.backcast_length, self.forecast_length)
+            # print(x.shape)
+            # print(y.shape)
+            # history = self.nbeats.train_on_batch(x, y)
+            self.nbeats.train_on_batch(x, y)
+
+            if step % self.plot_results == 0:
+                print(step)
+                if not os.path.exists('results'):
+                    os.makedirs('results')
+                self.nbeats.save('results/n_beats_model_20_' + str(step) + '.h5')
+                predictions = self.nbeats.predict(x_test)
+                residual = y_test - predictions
+                print(get_accuracy(y_test, predictions))
+                if get_accuracy(y_test, predictions)[0] < self.best_perf:
+                    self.best_perf = get_accuracy(y_test, predictions)[0]
+                    self.nbeats.save('results/n_beats_model_20_ongoing.h5')
+                print(predictions.shape)
+                print(residual.shape)
+                plt.figure()
+                plt.grid(True)
+                print(x.shape)
```

Comment:
Print not necessary. Or you can maybe add them on the same line :)

Reply:
Corrected